File: packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/types/model/FlutterAuthRule.kt
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

package com.amazonaws.amplify.amplify_datastore.types.model

import com.amplifyframework.core.model.AuthRule
import com.amplifyframework.core.model.AuthStrategy
import com.amplifyframework.core.model.ModelOperation

data class FlutterAuthRule(val map: Map<String, Any>) {

    private val authStrategy: AuthStrategy =
        stringToAuthStrategy(map["authStrategy"] as String)
    private val ownerField: String? = map["ownerField"] as String?
    private val identityClaim: String? = map["identityClaim"] as String?
    private val groupClaim: String? = map["groupClaim"] as String?
    private val groups: List<String>? = map["groups"] as List<String>?
    private val groupsField: String? = map["groupsField"] as String?
    private val operations: List<ModelOperation>? =
        (map["operations"] as List<String>?)?.map { stringToModelOperation(it) }
    private val authProvider: AuthStrategy.Provider =
        stringToAuthStrategyProvider(map["provider"] as String?) ?: authStrategy.defaultAuthProvider

    private fun stringToAuthStrategy(string: String): AuthStrategy {
        return when (string) {
            "OWNER" -> AuthStrategy.OWNER
            "GROUPS" -> AuthStrategy.GROUPS
            "PRIVATE" -> AuthStrategy.PRIVATE
            "PUBLIC" -> AuthStrategy.PUBLIC
            else -> throw Exception("stringToAuthStrategy - invalid string supplied: $string")
        }
    }

    private fun stringToModelOperation(string: String): ModelOperation {
        return when (string) {
            "CREATE" -> ModelOperation.CREATE
            "UPDATE" -> ModelOperation.UPDATE
            "DELETE" -> ModelOperation.DELETE
            "READ" -> ModelOperation.READ
            else -> throw Exception("stringToModelOperation - invalid string supplied: $string")
        }
    }

    private fun stringToAuthStrategyProvider(string: String?): AuthStrategy.Provider? {
        return when (string) {
            "APIKEY" -> AuthStrategy.Provider.API_KEY
            "OIDC" -> AuthStrategy.Provider.OIDC
            "IAM" -> AuthStrategy.Provider.IAM
            "USERPOOLS" -> AuthStrategy.Provider.USER_POOLS
            "FUNCTION" -> AuthStrategy.Provider.FUNCTION
            else -> null
        }
    }

    fun convertToNativeAuthRule(): AuthRule {

        val builder: AuthRule.Builder = AuthRule.builder()
            .authStrategy(authStrategy)

        builder.authProvider(authProvider)

        if (groups != null && groups.isNotEmpty()) {
            builder.groups(groups)
        }

        if (!ownerField.isNullOrEmpty()) {
            builder.ownerField(ownerField)
        }

        if (!identityClaim.isNullOrEmpty()) {
            builder.identityClaim(identityClaim)
        }

        if (!groupClaim.isNullOrEmpty()) {
            builder.groupClaim(groupClaim)
        }

        if (!groupsField.isNullOrEmpty()) {
            builder.groupsField(groupsField)
        }

        if (!operations.isNullOrEmpty()) {
            builder.operations(operations)
        }

        return builder.build()
    }
}

File: packages/amplify_datastore/ios/Classes/types/model/FlutterAuthRule.swift
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import Foundation
import Amplify

public struct FlutterAuthRule {
    private var authStrategy: String
    private let ownerField: String?
    private let identityClaim: String?
    private let groupClaim: String?
    private let groups: [String]?
    private let groupsField: String?
    private let provider: String?
    private var operations: [String]?

    init(serializedData: [String: Any]) throws {
        guard let authStrategy = serializedData["authStrategy"] as? String
        else {
            throw ModelSchemaError.parse(
                className: "FlutterAuthRule",
                fieldName: "authStrategy",
                desiredType: "String")
        }

        self.authStrategy = authStrategy

        self.ownerField = serializedData["ownerField"] as? String

        self.identityClaim = serializedData["identityClaim"] as? String

        self.groupClaim = serializedData["groupClaim"] as? String

        self.groups = serializedData["groups"] as? [String]

        self.groupsField = serializedData["groupsField"] as? String

        self.provider = serializedData["provider"] as? String

        self.operations = serializedData["operations"] as? [String]
    }

    private func stringToAuthStrategy(authStrategyString: String) -> AuthStrategy{
        switch authStrategyString {
            case "OWNER":
                return AuthStrategy.owner
            case "GROUPS":
                return AuthStrategy.groups
            case "PRIVATE":
                return AuthStrategy.private
            case "PUBLIC":
                return AuthStrategy.public
            default:
                preconditionFailure("Could not create a AuthStrategy from \(authStrategyString)")
        }
    }

    private func stringToAuthProvider(providerString: String?) -> AuthRuleProvider? {
        switch providerString {
        case "APIKEY":
            return AuthRuleProvider.apiKey
        case "OIDC":
            return AuthRuleProvider.oidc
        case "IAM":
            return AuthRuleProvider.iam
        case "USERPOOLS":
            return AuthRuleProvider.userPools
        case "FUNCTION":
            return AuthRuleProvider.function
        default:
            return nil
        }
    }

    private func stringToModelOperation(modelOperationString: String) -> ModelOperation{
        switch modelOperationString {
            case "CREATE":
                return ModelOperation.create
            case "UPDATE":
                return ModelOperation.update
            case "DELETE":
                return ModelOperation.delete
            case "READ":
                return ModelOperation.read
            default:
            preconditionFailure("Could not create a ModelOperation from \(modelOperationString)")
        }
    }

    public func convertToNativeAuthRule() -> AuthRule{
        return AuthRule(
            allow: stringToAuthStrategy(authStrategyString: authStrategy),
            ownerField: ownerField,
            identityClaim: identityClaim,
            groupClaim: groupClaim,
            groups: groups ?? [String](),
            groupsField: groupsField,
            provider: stringToAuthProvider(providerString: provider),
            operations: operations?.map {
                stringToModelOperation(modelOperationString: $0)
            } ?? [ModelOperation]()
        )
    }
}
